### ISSUE TITLE
Add riscv64 to (APT) repo

### ISSUE THREAD

**Issue by MichaIng** [webmin/webmin#2052]
It would be great if the `riscv64` architecture could be added to the [APT repo](https://download.webmin.com/download/repository/). The DEB [packages](https://download.webmin.com/download/repository/pool/contrib/w/webmin/) are compatible with "all" architectures anyway, and we tested them successfully on a VisionFive 2 RISC-V SBC with Debian Sid.

The same should be possible RPM, but we did not explicitly test it.

**Comment by swelljoe** [comment]
The apt repo should not claim to offer binaries it doesn't offer. What problem are you having installing `all` packages from the repo?

**Comment by MichaIng** [comment]
Since there is no `https://download.webmin.com/download/repository/dists/sarge/contrib/binary-riscv64/Packages` (and it is not listed in `https://download.webmin.com/download/repository/dists/sarge/Release`), a system with `dpkg --print-architecture` == `riscv64` is not able to use the repo with APT.

The packages themselves can be manually installed via `apt install ./package.deb` or `dpkg -i ./package.deb` indeed, since they have `Architecture: all`. If I am not mistaken, an alternative would be to merge all architectures in the repo into one `https://download.webmin.com/download/repository/dists/sarge/contrib/binary-all/Packages` (and link it in the `Release` file). Systems use "all" as fallback for their particular architecture, if available on the repo.

**Comment by swelljoe** [comment]
You've got something misconfigured there, I think, as you should be able to use an `all` apt repo, regardless of your architecture.

**Comment by swelljoe** [comment]
Er, wait...are the Webmin repos claiming to be architecture dependent? That shouldn't be.

**Comment by swelljoe** [comment]
Er, you're using the old (wrong) repos, though. Why? The docs provide a script to setup the right, current, repo. Which, as far as I know, is not architecture dependent (though I don't maintain it).

**Comment by swelljoe** [comment]
Oh, hell. The new one is wrong, too. It claims to provide a bunch of binaries that it doesn't provide (but not riscv64). And, it doesn't even seem to claim to provide `all`, which is what it _should_ be saying it provides. @Jamie, the arch of the Webmin repo should be `all` and none of that other stuff.

**Comment by swelljoe** [comment]
So, the solution here is not to _add_ riscv64 to the repo, but to remove all the other ones and add `all` (which is the only thing in that repo).

**Comment by MichaIng** [comment]
Yeah it is a little contradicting. The header show `Architecture: all i386 amd64`, but does not list/have a `binary-all/Packages`. APT seems to internally fall back to `binary-<own-arch>`, which works for all other architectures, but not for `riscv64`:
```console
# apt update
Hit:1 https://deb.debian.org/debian sid InRelease
Ign:2 https://download.webmin.com/download/repository sarge InRelease
Get:3 https://download.webmin.com/download/repository sarge Release [16.9 kB]
Get:4 https://download.webmin.com/download/repository sarge Release.gpg [181 B]
Fetched 17.0 kB in 3s (5952 B/s)
Reading package lists...
W: Skipping acquire of configured file 'contrib/binary-riscv64/Packages' as repository 'https://download.webmin.com/download/repository sarge InRelease' does not seem to provide it (sources.list entry misspelt?)
```

> So, the solution here is not to add riscv64 to the repo, but to remove all the other ones and add `all` (which is the only thing in that repo).

Theoretically the individual package architecture does not need to match the one listed in `Release`, as long as it is compatible, but I agree it is much cleaner to just have a single `all` architecture for the all `all` packages.

**Comment by swelljoe** [comment]
You're still looking at the wrong repo. You should use the new one.

**Comment by swelljoe** [comment]
But, the new one is also wrong. It's just wrong in different ways.

**Comment by swelljoe** [comment]
The docs for managing Debian repos, with any of the tools available, are simply horrible. So, it's not at all easy to get any of this right (even with aptly, which is probably the best tool currently available, though I don't think Jamie is using that...I assume he's using dpkg-scanpackages and a mess of other stuff as it used to be done). No matter what, it's hard to find clear, concise, information about the right way to do this.

But, I'm pretty confident a repo should not say it has packages for architectures it does not have packages for. And, `all` is for every architecture, and that's all the Webmin repo has.

**Comment by MichaIng** [comment]
Ah thanks for the hint, fount it: https://download.webmin.com/download/newkey/repository/dists/stable/Release
Compared to the old one, it is a little more consistent, as it contains all individual architectures in the header. But the basic issue is the same: No `binary-all/Packages`.

> But, I'm pretty confident a repo should not say it has binary packages for architectures it does not have packages for. And, `all` is for every architecture, and that's all the Webmin repo has.

Agreed.

**Comment by jcameron** [comment]
Yeah currently the repo has a bunch of per-architecture directories for ancient historical reasons. Sounds like the fix is to just create a binary-all directory?

**Comment by MichaIng** [comment]
Jep, something like this should do:
```sh
cp -a contrib/binary-{arm,riscv}64
```
and adding the hashes to the `Release` file (+ regenerating the `Release.gpg`), which is also just copy&pasting the lines as of course they are all the same for all architectures. But I guess there is a tool like `mini-dak` used to manage the APT repo, which surely allows to configure this properly.

**Comment by jcameron** [comment]
I added a `binary-all` directory to the repo .... please try installing again and let me know if that works.

**Comment by MichaIng** [comment]
@jcameron 
I see you added this to the old repository, not the new one? Tested with:
```console
root@visionfive2:~# cat /etc/apt/sources.list.d/webmin.list
deb https://download.webmin.com/download/repository/ sarge contrib
```
Interestingly, `apt update` shows the same warning, but the package installs regardless from `binary-all`:
```console
root@visionfive2:~# apt update
Hit:1 https://deb.debian.org/debian sid InRelease
Ign:2 https://download.webmin.com/download/repository sarge InRelease
Hit:3 https://download.webmin.com/download/repository sarge Release
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
All packages are up to date.
W: Skipping acquire of configured file 'contrib/binary-riscv64/Packages' as repository 'https://download.webmin.com/download/repository sarge InRelease' does not seem to provide it (sources.list entry misspelt?)
root@visionfive2:~# apt install webmin
Reading package lists...
Building dependency tree...
Reading state information...
The following additional packages will be installed:
  libauthen-pam-perl libgdbm-compat4 libgdbm6 libicu72 libio-pty-perl
  libnet-ssleay-perl libperl5.36 libxml2 perl perl-modules-5.36
  perl-openssl-defaults shared-mime-info
Suggested packages:
  gdbm-l10n perl-doc libterm-readline-gnu-perl | libterm-readline-perl-perl
  make libtap-harness-archive-perl debhelper
Recommended packages:
  netbase libdatetime-perl libdatetime-timezone-perl libdatetime-locale-perl
  libencode-detect-perl lynx
The following NEW packages will be installed:
  libauthen-pam-perl libgdbm-compat4 libgdbm6 libicu72 libio-pty-perl
  libnet-ssleay-perl libperl5.36 libxml2 perl perl-modules-5.36
  perl-openssl-defaults shared-mime-info webmin
0 upgraded, 13 newly installed, 0 to remove and 0 not upgraded.
Need to get 51.9 MB of archives.
After this operation, 281 MB of additional disk space will be used.
Get:1 https://deb.debian.org/debian sid/main riscv64 perl-modules-5.36 all 5.36.0-10 [2815 kB]
Get:2 https://deb.debian.org/debian sid/main riscv64 libgdbm6 riscv64 1.23-5 [73.2 kB]
Get:3 https://deb.debian.org/debian sid/main riscv64 libgdbm-compat4 riscv64 1.23-5 [48.3 kB]
Get:4 https://deb.debian.org/debian sid/main riscv64 libperl5.36 riscv64 5.36.0-10 [3827 kB]
Get:5 https://download.webmin.com/download/repository sarge/contrib all webmin all 2.105 [33.7 MB]
Get:6 https://deb.debian.org/debian sid/main riscv64 perl riscv64 5.36.0-10 [239 kB]
Get:7 https://deb.debian.org/debian sid/main riscv64 perl-openssl-defaults riscv64 7 [7404 B]
Get:8 https://deb.debian.org/debian sid/main riscv64 libnet-ssleay-perl riscv64 1.92-2 [315 kB]
Get:9 https://deb.debian.org/debian sid/main riscv64 libauthen-pam-perl riscv64 0.16-6 [25.6 kB]
Get:10 https://deb.debian.org/debian sid/main riscv64 libio-pty-perl riscv64 1:1.17-1 [35.1 kB]
Get:11 https://deb.debian.org/debian sid/main riscv64 libicu72 riscv64 72.1-4 [9375 kB]
Get:12 https://deb.debian.org/debian sid/main riscv64 libxml2 riscv64 2.9.14+dfsg-1.3+b1 [706 kB]
Get:13 https://deb.debian.org/debian sid/main riscv64 shared-mime-info riscv64 2.4-1 [750 kB]
...
```

**Comment by jcameron** [comment]
Hmm, I wonder if this means that APT requires that all repositories contain all architectures?

**Comment by MichaIng** [comment]
As I could not find any info about APT repositories in general in the [Debian policies](https://www.debian.org/doc/debian-policy/), I just asked in the developer mailing list: https://lists.debian.org/debian-devel/2023/12/msg00023.html

**Comment by jcameron** [comment]
Ok, we'll add a `riscv64` architecture explicitly in the next release.
